wsgi.py

```diff
@@ -6,255 +6,349 @@
 import re
 import os
 from flask import Flask, redirect, url_for, request, Response
-#from flask_restful import reqparse
+from flask_httpauth import HTTPBasicAuth
+
+# from flask_restful import reqparse
 
 import sys
 
-from openshift_rolebindings import *
-from openshift_project import *
-from openshift_identity import *
-from openshift_user import *
+from openshift import *
 
 application = Flask(__name__)
+auth = HTTPBasicAuth()
 
-if __name__ != '__main__':
-    gunicorn_logger = logging.getLogger('gunicorn.error')
+if __name__ != "__main__":
+    gunicorn_logger = logging.getLogger("gunicorn.error")
     application.logger.handlers = gunicorn_logger.handlers
     application.logger.setLevel(gunicorn_logger.level)
 
 
-def get_user_token():
-    with open('/var/run/secrets/kubernetes.io/serviceaccount/token', 'r') as file:
+def get_openshift():
+    version = os.environ["OPENSHIFT_VERSION"]
+    url = os.environ["OPENSHIFT_URL"]
+    with open("/var/run/secrets/kubernetes.io/serviceaccount/token", "r") as file:
         token = file.read()
-        return token
-    return ""
-
-def get_token_and_url():
-    token = get_user_token()
-    openshift_url = os.environ["openshift_url"]
-    return (token, openshift_url)
-
-@application.route("/users/<user_name>/projects/<project_name>/roles/<role>", methods=['GET'])
+        if version == '3':
+            shift = openshift_3_x(url, token, application.logger)
+            application.logger.info("using Openshift ver 3")
+        else:
+            shift = openshift_4_x(url, token, application.logger)
+            application.logger.info("using Openshift ver 4")
+        return shift
+    application.logger.info("Unable to open service account token file, shift not initialized")
+
+
+@auth.verify_password
+def verify_password(username, password):
+    with open("/app/auth/users", "r") as f:
+        user_str = f.read()
+    if(user_str is not None and user_str.len() > 0):
```

Comment:
Strings don't have a `.len` attribute. You want `len(user_str)`,  but in fact since an empty string evaluates as `false` in a boolean context, you can just write:

```
if user_str:
```

Reply:
ok, have changed it.  Just the simple observation that Python allows this loose of syntax to go without breaking leads me to believe that it is broken on a fundamental level.

Reply:
That's a fundamental flaw in any language with dynamic typing. (and it's a semantic issue, not a syntax one - looking at line 45 in isolation, "user_str" *could* have a method named "len" - it's just that in this case it doesn't)

If we wanted to go back and add static typing to all our code we could catch these cases - mypy flags this as:
  error: "str" has not attribute "len"

I'm not actually suggesting that we do that, just pointing out that it's an option.

Reply:
actually pylint runs mypy, and flags other things like that.